File: client/src/main/java/heast/client/network/ClientLoginHandler.java
package heast.client.network;

import heast.core.logging.IO;
import heast.core.network.NetworkState;
import heast.core.network.packets.c2s.auth.LoginC2SPacket;
import heast.core.network.packets.c2s.login.KeyC2SPacket;
import heast.core.network.packets.s2c.login.HelloS2CPacket;
import heast.core.network.pipeline.ClientConnection;
import heast.core.network.listeners.ClientLoginListener;
import heast.core.security.AES;

import javax.crypto.SecretKey;
import java.security.KeyFactory;
import java.security.NoSuchAlgorithmException;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.X509EncodedKeySpec;

public final class ClientLoginHandler implements ClientLoginListener {

    private final ClientConnection connection;
    private SecretKey secretKey;

    public ClientLoginHandler(ClientConnection connection) {
        this.connection = connection;
    }

    /**
     * Called when the server acknowledges the client's connection.
     * @param packet The packet containing the server's public RSA key.
     */
    @Override
    public void onHello(HelloS2CPacket packet) {
        try {
            var encodedKeySpec = new X509EncodedKeySpec(packet.getKey());
            var publicKey = KeyFactory.getInstance("RSA")
                .generatePublic(encodedKeySpec);

            secretKey = AES.generateKey();
            connection.send(new KeyC2SPacket(secretKey, publicKey), () -> {
                connection.enableEncryption(secretKey);
                IO.info.println("Enabled encryption.");
            });
        } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
            IO.error.println("Failed to construct RSA key.");
        }
    }

    /**
     * Called when the client successfully authenticates with the server.
     */
    @Override
    public void onSuccess() {
        connection.setState(NetworkState.AUTH);
        connection.setListener(new ClientAuthHandler(connection));
    }
}

File: Client/Network/ClientLoginHandler.cs
using System;
using System.Security.Cryptography;
using Core.Network;
using Core.Network.Codecs;
using Core.Network.Listeners;
using Core.Network.Packets.C2S;
using Core.Network.Packets.S2C;

namespace Client.Network;

public class ClientLoginHandler : IClientLoginListener
{
	public ClientLoginHandler(ClientConnection ctx)
	{
		Ctx = ctx;
	}

	private ClientConnection Ctx { get; }

	private Aes? KeyPair { get; set; }

	public void OnHello(HelloS2CPacket packet)
	{
		var key = new byte[128];
		var iv = new byte[128];
		KeyPair = Aes.Create();
		{
			KeyPair.Mode = CipherMode.CFB;
			KeyPair.Padding = PaddingMode.PKCS7;
			RandomNumberGenerator.Fill(key);
			RandomNumberGenerator.Fill(iv);
		}

		var encryptedKey = new Span<byte>();
		var encryptedIv = new Span<byte>();
		RSACryptoServiceProvider rsa = new(4096);
		{
			rsa.ImportRSAPublicKey(packet.Key, out _);
			var success = rsa.TryEncrypt(
				key, encryptedKey, RSAEncryptionPadding.OaepSHA256, out _);
			success &= rsa.TryEncrypt(
				iv, encryptedIv, RSAEncryptionPadding.OaepSHA256, out _);
			if (!success)
			{
				const string error = "Failed to encrypt key";
				Ctx.Send(new ErrorC2SPacket(Error.InvalidKey, error));
				throw new(error);
			}
		}

		Ctx.Send(new KeyC2SPacket(
			encryptedKey.ToArray(),
			encryptedIv.ToArray()));
	}

	public void OnSuccess()
	{
		Ctx.EnableEncryption(KeyPair!);
		Ctx.State = NetworkState.Auth;
		Ctx.Listener = new ClientAuthHandler(Ctx);
	}

	public void OnError(ErrorS2CPacket packet)
	{
		throw new NotImplementedException();
	}
}